controllers/webhook/runcompletion_feed.go

```diff
@@ -0,0 +1,111 @@
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	config "github.com/sky-uk/kfp-operator/apis/config/v1alpha5"
+	"github.com/sky-uk/kfp-operator/apis/pipelines"
+	"io"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type EventData struct {
+	// Header is the http request header
+	Header http.Header `json:"header"`
+	// Body is http request body
+	Body json.RawMessage `json:"body"`
+}
+
+type RunCompletionFeed struct {
+	ctx       context.Context
+	upstreams []UpstreamService
+}
+
+func NewRunCompletionFeed(ctx context.Context, endpoints []config.Endpoint) RunCompletionFeed {
+	upstreams := pipelines.Map(endpoints, func(endpoint config.Endpoint) UpstreamService {
+		return NewHttpWebhook(endpoint)
+	})
+	return RunCompletionFeed{
+		ctx:       ctx,
+		upstreams: upstreams,
+	}
+}
+
+func getRequestBody(ctx context.Context, request *http.Request) ([]byte, error) {
+	logger := log.FromContext(ctx)
+
+	if request.Body == nil {
+		return nil, errors.New("request body is nil")
+	}
+
+	// Ensure that the request body is closed after reading
+	defer func(body io.ReadCloser) {
+		err := body.Close()
+		if err != nil {
+			logger.Error(err, "Failed to close request body")
+		}
+	}(request.Body)
+
+	body, err := io.ReadAll(request.Body)
+	request.Body = io.NopCloser(bytes.NewBuffer(body))
+	if err != nil {
+		return nil, fmt.Errorf("failed to read request body, %w", err)
+	} else if len(body) == 0 {
+		return nil, errors.New("request body is empty")
+	}
+	return body, nil
+}
+
+func extractEventData(ctx context.Context, request *http.Request) (*EventData, error) {
+	body, err := getRequestBody(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	rawMessage := json.RawMessage(body)
+	return &EventData{
+		Header: request.Header,
+		Body:   rawMessage,
+	}, nil
+}
+
+func (rcf RunCompletionFeed) handleEvent(response http.ResponseWriter, request *http.Request) {
+	logger := log.FromContext(rcf.ctx)
+	switch request.Method {
+	case http.MethodPost:
+		if request.Header.Get("Content-Type") != "application/json" {
```

Comment:
Worth putting `Content-Type` and `application/json` as global constants somewhere since they are re-used in multiple places?

Reply:
well you would think go would have already provided that in base library wouldnt you 🐒

Reply:
I have done anyway, consts at top of run completion feed file, yes would have been nice to not just be strings and ints but...

Reply:
See [d626eca](https://github.com/sky-uk/kfp-operator/pull/368/commits/d626ecacae21d54eb78ccd9c1dcb1075cbef2215)